### ISSUE TITLE
Casting error

### ISSUE THREAD

**Issue by babu-akhil** [siddhartha-gadgil/ProvingGround#244]
```
indfn = AxB.induc(A.Var :-> (B.Var :-> C))
    recfn = AxB.rec(C)
    indfn == recfn
```

returns error

```
java.lang.ClassCastException: provingground.HoTT$PiSymbolicFunc cannot be cast to provingground.HoTT$Func
  provingground.HoTT$ProdTyp$RecFn.subs(HoTT.scala:1034)
  provingground.HoTT$ProdTyp$RecFn.subs(HoTT.scala:1014)
  provingground.HoTT$Subs.replace(HoTT.scala:212)
  provingground.HoTT$Subs.replace$(HoTT.scala:183)
  provingground.HoTT$ProdTyp$RecFn.replace(HoTT.scala:1014)
  provingground.HoTT$LambdaLike.equals(HoTT.scala:1888)
  provingground.HoTT$LambdaLike.equals$(HoTT.scala:1886)
  provingground.HoTT$LambdaFixed.equals(HoTT.scala:1946)
  ammonite.$sess.cmd9$.<init>(cmd9.sc:1)
  ammonite.$sess.cmd9$.<clinit>(cmd9.sc)
```

**Comment by siddhartha-gadgil** [comment]
I am working on this. Indeed this is a more basic error.
```scala
import provingground._ , interface._, HoTT._, learning._
val A = Type.sym
val B = Type.sym
val a = A.sym
val b = B.sym
val C = Type.sym
val p = (A ->: (B ->: C)).Var
val q = (a ~>: (b ~>: C)).Var
p.replace(p, q)
```
gives error
```scala
java.lang.ClassCastException: provingground.HoTT$PiSymbolicFunc cannot be cast to provingground.HoTT$Func
  ammonite.$sess.cmd20$.<init>(cmd20.sc:1)
  ammonite.$sess.cmd20$.<clinit>(cmd20.sc)
```

**Comment by siddhartha-gadgil** [comment]
This is a complex issue involving type erasure, but is going to get a quick-fix. Here are details recorded before applying this:

* firstly, there is an assumption that when we substitute `x` by `y` in a term `t` with scala type `U`, we generally get a result with type `U`. This is clearly nonsense for arbitrary substitutions, e.g. when `x = t` and `y` has a different type.
* So we only make substitutions that make good sense.
* One place we make them is in comparing functions defined as lambdas, say `x1 :-> y1` and `x2 :-> y2`. In this case, we first check that the HoTT types of `x1` and `x2` are equal, and only then substitute `x1` by `x2` in `y1` and compare the result to `y2`.
* However, in some cases the scala types may differ even if the HoTT types are equal, leading to cast errors. This happens for example because constant Pi-types are equal to function types.
* Unfortunately type erasure means that one cannot check if the casting worked fine before returning a result (the JVM declares it worked, and fails later).
* The quick-fix is to wrap the replacement based equality check in a `Try` and declare not equal if there is an error.

**Comment by babu-akhil** [comment]
If i understand correctly, the Scala type of constant Pi-type is not the same as function type. But cant we have pure function type as a sub-type of dependent function type(since all pure functions are dependent functions with a constant type family), and hence we should be able to check equality between them?
 say we have a pure function `f1` and dependent function `f2`, then ` f1 == f2` could be checked by converting `f1` to a dependent function  'depf1' with constant family and then checking `depf1 = f2`

**Comment by siddhartha-gadgil** [comment]
Yes, and we do check equalities in this manner for example for lambdas,
recursive functions etc. The issue is that substituting a pure function by
a dependent one can give an error violating the assumption that
substitutions preserve types. We can add try to add special rules, say to
the `replace` function to avoid such errors (many special rules have been
added over time). But it may not be worth the effort unless we have wrong
behaviour in some cases.

**Comment by siddhartha-gadgil** [comment]
Fixed in a cleaner way in 1a2fa81ab33b2a3ee97e1b60
